conmon-rs/server/src/child_reaper.rs

```diff
@@ -0,0 +1,124 @@
+//! Child process reaping and management.
+
+use anyhow::{bail, Result};
+use getset::Getters;
+use log::{debug, error};
+use nix::sys::{
+    signal::{kill, Signal, SIGCHLD},
+    signalfd::{siginfo, SfdFlags, SigSet, SignalFd},
+};
+use nix::unistd::Pid;
+use std::path::PathBuf;
+use std::thread;
+use std::{collections::HashMap, fs::File, io::Write, sync::Arc, sync::Mutex};
+
+impl ChildReaper {
+    pub fn init_reaper(&self) -> Result<()> {
+        let mut mask = SigSet::empty();
+        mask.add(SIGCHLD);
```

Comment:
BTW, see https://github.com/containers/containers-image-proxy-rs/pull/13 in which I discovered that tokio wants to install its own `SIGCHLD` handler.  And I think using `signalfd()` here would conflict with that.  

This is all probably fine but be aware that things will go wrong if you end up spawning a process [using tokio](https://docs.rs/tokio/1.14.0/tokio/process/struct.Command.html).

(The long term fix is to use pidfd - which, actually why aren't you using pidfd here?)

Reply:
does rhel 8 support pidfds? I didn't implement with them because I didn't think it did

Reply:
Yeah you're right, it doesn't :cry:

Reply:
OK I skimmed through the systemd sources here https://github.com/systemd/systemd/blob/361bf5fb770d4180eb5407f592a43c8f288c750f/src/libsystemd/sd-event/sd-event.c#L1312

Since I was curious if this thread spawning was really a best practice.  There's levels of nontrivial logic there.

But, systemd goes to great lengths to minimize its use of threads (particularly in pid1).  See e.g. https://sourceware.org/bugzilla/show_bug.cgi?id=26371

But in Rust land threads are the default and we have memory safety etc.

Would it work to just have a thread calling `waitpid(-1, ...)` in a loop?  That would seem to be dramatically simpler.

Reply:
good idea, this is a couple of lines simpler